### PR TITLE
[FIX] survey: wait for user inputs to be loaded in tour

### PR DESCRIPTION
**Pull request by flch-odoo**
In this tour, we trigger the loading of participants input immediately
after the widget's start.

This is however a chain of promises, and it sometimes happened that the
tour proceeded to the next step between these, i.e., before the results
were fetched.

Task-3284627

### CODE REVIEW

File: addons/survey/static/tests/tours/survey_tour_session_manage.js [hunk 1]
```diff
@@ -20,6 +20,11 @@ var getChartData = function () {
         return chartData;
     }
 
+    // Wait for the chart to be fully loaded (after animations)
+    await new Promise((resolve) => {
```
Comment (abd-msyukyu-odoo) [inline addons/survey/static/tests/tours/survey_tour_session_manage.js]:
```js
    await surveyManagePublicWidget.resultsChart.fullyRendered
```

**Review by abd-msyukyu-odoo — COMMENTED** [review]
Hello! Just wanted to send you what I was talking about (solution without `setTimeout`) ;p

**Comment by flch-odoo** [comment]
Thanks for the suggestion, @abd-msyukyu-odoo ! I have worked with it, implemented a version with it, and learned a few things along the way. Ultimately, I now think the solution was elsewhere in this case (because of some tour-specific "magic", it would have become much more complex for the same result); but I won't forget this :muscle:.

File: addons/survey/static/tests/tours/survey_tour_session_manage.js [hunk 2]
```diff
@@ -31,47 +34,99 @@ var getChartData = function () {
...
 /**
  * The check is based on the specific opacity (0.2) and color of incorrect answers.
  */
-var isIncorrectAnswer = function (answer) {
+const isIncorrectAnswer = function (answer) {
     return answer.backgroundColor.includes('0.2') &&
         answer.labelColor === WRONG_ANSWER_COLOR;
 };
 
+const _checkAnswersCount = function (chartData, expectedCount) {
```
Comment (awa-odoo) [inline addons/survey/static/tests/tours/survey_tour_session_manage.js:78]: Would not keep an underscore at the beginning to match syntax of other functions.

We don't really do "private functions" anymore.
I would also apply this to your constant dictionaries.

File: addons/survey/static/tests/tours/survey_tour_session_manage.js [hunk 3]
```diff
@@ -9,10 +12,10 @@ import surveySessionTools from "survey.session_tour_tools";
  * This helper method returns the chart data (Chartjs framework specific) in the following structure:
  * [{ value, backgroundColor, labelColor }]
  */
-var getChartData = function () {
-    var chartData = [];
-    var rootWidget = odoo.__DEBUG__.services['root.widget'];
-    var surveyManagePublicWidget = rootWidget.publicWidgets.find(function (widget) {
+const getChartData = function () {
```
Comment (awa-odoo) [inline addons/survey/static/tests/tours/survey_tour_session_manage.js:15]: If we're switching to "const" we could also switch to arrow functions?

File: addons/survey/static/tests/tours/survey_tour_session_manage.js [hunk 4]
```diff
@@ -31,47 +34,99 @@ var getChartData = function () {
...
+    }
+};
+
+const _IS_ASPECT_ANSWER_DATA = {
+    "correct": [isCorrectAnswer, 'should be shown as "correct"!'],
+    "incorrect": [isIncorrectAnswer, 'should be shown as "incorrect"!'],
+    "regular": [isRegularAnswer, 'should not be shown as "correct" or "incorrect"!'],
+    "regularScored": [isRegularAnswer, "correctness shouldn't be shown!"],
+};
+const _INDEX_TO_LABEL = {
```
Comment (awa-odoo) [inline addons/survey/static/tests/tours/survey_tour_session_manage.js:96]: Missing blank space before this constant (such review).

File: addons/survey/static/tests/tours/survey_tour_session_manage.js [hunk 5]
```diff
@@ -31,47 +34,99 @@ var getChartData = function () {
...
+    if (chartData.find(answerData => answerData !== 0).length) {
+        throw new TourError('Chart data should all be 0!');
+    }
+};
+
+const _IS_ASPECT_ANSWER_DATA = {
+    "correct": [isCorrectAnswer, 'should be shown as "correct"!'],
+    "incorrect": [isIncorrectAnswer, 'should be shown as "incorrect"!'],
+    "regular": [isRegularAnswer, 'should not be shown as "correct" or "incorrect"!'],
+    "regularScored": [isRegularAnswer, "correctness shouldn't be shown!"],
```
Comment (awa-odoo) [inline addons/survey/static/tests/tours/survey_tour_session_manage.js:94]: Not sure we need this, keeping "regular" only should be enough.
"regularScored" does not really qualify as an aspect anyway as it does the same check, only the error message slightly differs.

File: addons/survey/static/tests/tours/survey_tour_session_manage.js [hunk 6]
```diff
@@ -31,47 +34,99 @@ var getChartData = function () {
...
+    }
+};
+
+const _checkAllZeros = function (chartData) {
+    if (chartData.find(answerData => answerData !== 0).length) {
+        throw new TourError('Chart data should all be 0!');
+    }
+};
+
+const _IS_ASPECT_ANSWER_DATA = {
```
Comment (awa-odoo) [inline addons/survey/static/tests/tours/survey_tour_session_manage.js:90]: Not convinced about "aspect", took me a while to figure out what you meant.

Maybe a good old "type" ? "ANSWER_TYPE_CHECK_MAPPING" ?

We could also simply do a if/else in the "checkAnswers" method, as I'm not sure the code reduction is worth the added reading complexity, but I will let you have the final word :)

File: addons/survey/static/tests/tours/survey_tour_session_manage.js [hunk 7]
```diff
@@ -31,47 +34,99 @@ var getChartData = function () {
...
+ * @param {{value: number, aspect: "correct" | "incorrect" | "regular" | "regularScored"}[]} expectedAnswersData with:
+ *  `value`: expected answers count (number of user inputs)
+ *  `aspect`: Appearance of the answers count. Use `regularScored` for scored question
+ *    (when the correctness should not be shown) to improve the error message
+ */
+const checkAnswers = function (chartData, expectedAnswersData) {
+    _checkAnswersCount(chartData, expectedAnswersData.length);
+
+    zip(chartData, expectedAnswersData).forEach(([actual, expected], answerIndex) => {
+        const [isAspectAnswer, aspectErrorMessage] = _IS_ASPECT_ANSWER_DATA[expected.aspect];
```
Comment (awa-odoo) [inline addons/survey/static/tests/tours/survey_tour_session_manage.js:116]: Would do "[isAnswerType, errorMessage]" to match other suggested naming.

File: addons/survey/static/tests/tours/survey_tour_session_manage.js [hunk 8]
```diff
@@ -31,47 +34,99 @@ var getChartData = function () {
...
+        const questionLabel = _INDEX_TO_LABEL[answerIndex];
+        if (!isAspectAnswer(actual)) {
+            throw new Error(`${questionLabel} answer ${aspectErrorMessage}!`);
```
Comment (awa-odoo) [inline addons/survey/static/tests/tours/survey_tour_session_manage.js:119]: Usually not a fan of messages crafted in multiple parts but I guess it's ok for a test as it helps factorizing some bits.

File: addons/survey/static/tests/tours/survey_tour_session_manage.js [hunk 9]
```diff
@@ -31,47 +34,99 @@ var getChartData = function () {
...
+        }
+        if (actual.value !== expected.value) {
+            throw new Error(expected.value ?
```
Comment (awa-odoo) [inline addons/survey/static/tests/tours/survey_tour_session_manage.js:122]: Could check specifically that it's different from 0.

File: addons/survey/static/tests/tours/survey_tour_session_manage.js [hunk 10]
```diff
@@ -31,47 +34,99 @@ var getChartData = function () {
...
+ * @param {object} chartData Object returned by `getChartData`.
+ * @param {{value: number, aspect: "correct" | "incorrect" | "regular" | "regularScored"}[]} expectedAnswersData with:
+ *  `value`: expected answers count (number of user inputs)
+ *  `aspect`: Appearance of the answers count. Use `regularScored` for scored question
+ *    (when the correctness should not be shown) to improve the error message
+ */
+const checkAnswers = function (chartData, expectedAnswersData) {
+    _checkAnswersCount(chartData, expectedAnswersData.length);
+
+    zip(chartData, expectedAnswersData).forEach(([actual, expected], answerIndex) => {
```
Comment (awa-odoo) [inline addons/survey/static/tests/tours/survey_tour_session_manage.js:115]: Maybe "answer" and "expectedAnswer" ? "actual" does not mean much taken out of context.

File: addons/survey/static/tests/tours/survey_tour_session_manage.js [hunk 11]
```diff
@@ -175,197 +230,72 @@ registry.category("web_tour.tours").add('test_survey_session_manage_tour', {
     run: nextScreen
 }, {
     trigger: 'h1:contains("Regular Simple Choice")',
+    extra_trigger: '.o_survey_session_progress_small[style*="width: 100%"]',
```
Comment (awa-odoo) [inline addons/survey/static/tests/tours/survey_tour_session_manage.js:233]: As this is fairly specific, maybe add a small code comment that is similar to what you have in the commit message?

File: addons/survey/static/tests/tours/survey_tour_session_manage.js [hunk 12]
```diff
@@ -175,197 +230,72 @@ registry.category("web_tour.tours").add('test_survey_session_manage_tour', {
...
-            console.error(
-                'Fourth answer should be picked by no users and its correctness should not be shown yet!'
-            );
-            return;
-        }
-
+    }
+}, {
+    trigger: 'h1:contains("Scored Simple Choice")',
+    extra_trigger: '.o_survey_session_navigation_next_label:contains("Show Correct Answer(s)")',
```
Comment (awa-odoo) [inline addons/survey/static/tests/tours/survey_tour_session_manage.js:253]: Same idea for those, would add a small code comment.

File: addons/survey/static/tests/tours/survey_tour_session_manage.js [hunk 13]
```diff
@@ -1,6 +1,9 @@
 /** @odoo-module **/
 
 import { registry } from "@web/core/registry";
+import { TourError } from "@web_tour/tour_service/tour_utils";
+import { zip } from "@web/core/utils/arrays";
```
Comment (awa-odoo) [inline addons/survey/static/tests/tours/survey_tour_session_manage.js:5]: TIL we have a "zip" in JS :smile:

**Review by awa-odoo — COMMENTED** [review]
Hello @flch-odoo 

Finally took some time to review this, sorry for the delay :cold_face: 

It looks like the final solution is different from what we discussed, apparently you found a way to "synchronize" the steps by using specific extra_trigger.
Which is probably more simple and is "the tour way" so that's great.

I left some comments, mostly about naming and styling.

Thanks again for your investigation on this, I know it was not easy, and also for the nice bonus refactoring.

Cheers!

**Comment by awa-odoo** [comment]
@flch-odoo side-note: runbot seems very mad, probably needs a good rebase

**Review by flch-odoo — COMMENTED** [review]
Thanks for the review,  @awa-odoo!
I did replaced the mapping with if/else, and I felt there was no real need for different "check" functions anymore => Simpler code and a few less lines again.
The diff is a bit shaken up, but all should be correctly ordered.

**Comment by awa-odoo** [comment]
@robodoo r+

**Pull request closed without merging**